rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,545 @@
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits (REFERENCE ONLY!)
+
+**These traits are provided as reference to illustrate how these primitive shape types might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.**
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+impl Meshable for Direction {}
+
+struct Plane {
+  point: Point,
+  normal: Direction,
+}
+impl Meshable for Plane {}
+
+/// Differentiates a line from a ray, where a line is infinite and a ray is directional half-line, although their underlying representation is the same.
+struct Ray(
+  point: Point, 
+  direction: Direction,
+);
+impl Meshable for Ray {}
+
+// Line types
+
+/// Unbounded line in 3D space with directionality
+struct Line { 
+  point: Point, 
+  direction: Direction,
+}
+impl Meshable for Line {}
+
+/// A line segment bounded by two points
+struct LineSegment { 
+  start: Point, 
+  end: Point,
+}
+impl Meshable for LineSegment {}
+
+/// A line drawn along a path of points
+struct PolyLine {
+  points: Vec<Point>,
+}
+impl Meshable for PolyLine {}
+
+struct Triangle([Point; 3]);
+impl Meshable for Triangle {}
+impl Collider for Triangle {}
+
+struct Quad([Point; 4]);
+impl Meshable for Quad {}
+impl Collider for Quad {}
+
+/// Sphere types
+
+struct Sphere {
+  radius: f32,
+}
+impl Meshable for Sphere {}
+
+/* REFERENCE ONLY
+struct BoundingSphere {
+  sphere: Sphere,
+  translation: Vec3,
+}
+impl Meshable for BoundingSphere {}
+impl Bounding for BoundingSphere {}
+
+struct SphereCollider {
+  sphere: Sphere,
+  translation: Vec3,
+}
+impl Meshable for SphereCollider {}
+impl Collider for SphereCollider {}
+*/
+
+// Box Types
+
+struct Box {
+  half_extents: Vec3,
```

Comment:
half_extents are less user friendly (users generally think in width / height / depth). so users need to mentally (or in code) divide their desired extents when producing a box in code. however most algorithms prefer half-extents (because they offset from the center). The current impl optimizes for the algorithms. This might be the right call, but its worth discussing rationale and use cases.

Reply:
IMO internal storage should always be in the algorithm-friendly forms. User-friendly initialization can be handled in the form of various builder methods.

We use the same approach with `Timer`, where we have a `from_seconds` method even though the internal storage is a (much more correct) `Duration`.

Reply:
> The current impl optimizes for the algorithms. This might be the right call, but its worth discussing rationale and use cases.

I think this is a pretty fair generalization of my motivations. As Alice said, we can add constructors that are more intuitive independently of the in-memory representation. I think this is a case where we can have nice things!

Reply:
I find max/min Vec3's efficient for computing intersections and unions of bounding boxes, because you can use min and max per component to compute the result.

Also consider representation of an empty box, which is different from a point (half_extents = Vec3::ZERO), because it has no position. This is required to implement a monoid, where the empty box is the identity and the operation is the minimum box containing the union (i.e. bounding box). The bounding box of two points (zero sized boxes) at (0, 0, 0) and (1, 1, 1) would be a unit box, while the bounding box of an empty box and a point at (1, 1, 1) would be just the point.

This is useful to implement a bounding volume hierarchy from bounding boxes, because trees also form a monoid, such that an empty tree has a bound of the empty box.

Reply:
I also prefer min/max to centre + half extents but I guess it depends on the usage. I expect methods would be needed to get the centre, min, max, extents / half extents whatever the internal representation is.

Reply:
I think there are solid reasons to have both representations. I think I'm biased toward "user friendly" types having the "good names" (ex: Box would be defined by a public "extents" field), then we can add more specialized types that people with more specialized use cases can reach for (ex: MinMaxBox).

I think theres also strong argument to be made for putting all "mesh-compatible shapes" into their own module for easy discovery. I'm guessing users trying to create meshes dont wont to go through a trial and error processes of seeing if a type is convertible to a mesh.

```rust
/// user friendly mesh descriptions. maybe continues to live in bevy_render
/// maybe these have Mesh suffixes? CircleMesh, CubeMesh, etc
pub mod mesh_shapes { 
  Circle, Cube, Box, UvSphere, ...
}

/// math-friendly descriptions of shapes. lives in bevy_math or some new crate
pub mod math_shapes {
  RadiusCircle, DiameterCircle, ExtentBox, HalfExtentBox, MinMaxBox, ...
} 
```

Reply:
> I think I'm biased toward "user friendly" types having the "good names"

This RFC claims to target performance for intensive usage such as ray casting; this should dictate the type layout (and use SIMD types).